src/main/java/org/apache/maven/plugins/help/EffectivePomMojo.java

```diff
@@ -298,8 +300,49 @@ else if ( e.getTargetException() instanceof RuntimeException )
 
     private static String toString( InputLocation location )
     {
-        InputSource source = location.getSource();
+        String source = toString( location.getSource() );
+        String hierarchy = toHierarchyString( location.getSource() );
+
+        return '}' + source + ( ( location.getLineNumber() >= 0 ) ? ", line " + location.getLineNumber() : "" )
+                + hierarchy + ' ';
+    }
+
+    private static String toHierarchyString( InputSource source )
+    {
+        StringBuilder sb = new StringBuilder();
+
+        try
+        {
+            // InputSource#getImportedBy() added in maven-model 4.0.0-alpha-1
+            Method getImportedBy = InputSource.class.getDeclaredMethod( "getImportedBy" );
+            InputSource importedBy = (InputSource) getImportedBy.invoke( source );
+
+            while ( importedBy != null )
+            {
+                sb.append( " via " ).append( toString( importedBy ) );
+                importedBy = (InputSource) getImportedBy.invoke( importedBy );
+            }
 
+            return sb.toString();
+        }
+        catch ( NoSuchMethodException | SecurityException | IllegalArgumentException | IllegalAccessException
+                | ClassCastException e )
+        {
+            // Just continue without hierarchy
```

Comment:
Please add a debug log statement explaining that we will not show "imported by" because we're probably running Maven pre-4.

Reply:
Will do.

Reply:
This is actually just the case for the `NoSuchMethodException`. The other exceptions should never be thrown and reflect other issues. I would designate these "unexpected".

Reply:
Agreed.